src/Polly/CircuitBreaker/CircuitStateController.cs

```diff
@@ -126,14 +133,32 @@ protected void ResetInternal_NeedsLock(Context context)
 
         protected bool PermitHalfOpenCircuitTest()
         {
-            long currentlyBlockedUntil = _blockedTill;
-            if (SystemClock.UtcNow().Ticks >= currentlyBlockedUntil)
+            long initialBlockedTill = _blockedTill;
+            if (SystemClock.UtcNow().Ticks < initialBlockedTill)
             {
-                // It's time to permit a / another trial call in the half-open state ...
-                // ... but to prevent race conditions/multiple calls, we have to ensure only _one_ thread wins the race to own this next call.
-                return Interlocked.CompareExchange(ref _blockedTill, SystemClock.UtcNow().Ticks + _durationOfBreak.Ticks, currentlyBlockedUntil) == currentlyBlockedUntil;
+                return false;
             }
-            return false;
+
+            // It's time to permit a / another trial call in the half-open state ...
+            // ... but to prevent race conditions/multiple calls, we have to ensure only _one_ thread wins the race to own this next call.
+            // If "we" are the thread that wins that races, then we also want to update the BlockedTill date, so that no other thread / call tries to execute before this Test has reached *it's* conclusion.
+            //
+            // So combine doing those two things into an Interlocked.CompareExchange call.
+            // Attempt to update _blockedTill. Then examine the return value to determine whether someone else had beaten us to the punch.
+            // If not, then that tells us that we won the race, and we can permit the Circuit Test.
+            // If someone had already modified _blockedTil, then we lost the race and thus aren't going to be doing the Test.
+            //
+            // Note that if the Circuit Test is a failure, it will re-configure the Break limit, using the updated number of Test failures.
+            // Which means that it doesn't really matter what delay we put in here as long as no other attempts get made before this Test completes.
+            // This is helpful, since we don't have convenient access to the number of Test Failures here, and thus can't calculate the *correct* next delay.
+            //
+            // Lastly note that the conceptual edge case of _blockedTil being equal to DateTime.MaxValue.Ticks, is not a concern, since in that case the
+            // if-check above will not have passed, so we won't have gotten here in the first place.
```

Comment:
Fair points.

So I've started by factoring out that "extend it, capping at the limit, if necessary":

```
        private void BreakFor_NeedsLock(TimeSpan durationOfBreak, Context context)
        {
            var (breakEnd, breakLength) = ExtensionUpToMaxValue(SystemClock.UtcNow(), durationOfBreak);

            _currentBlockDuration = breakLength;
            _blockedUntil = breakEnd.Ticks;

            var transitionedState = _circuitState;
            _circuitState = CircuitState.Open;

            _onBreak(_lastOutcome, transitionedState, durationOfBreak, context);
        }

        private Tuple<DateTime, TimeSpan> ExtensionUpToMaxValue(DateTime startPoint, TimeSpan proposedExtension, DateTime? maxDateTime = null)
        {
            var maxPossibleExtension = ((maxDateTime ?? DateTime.MaxValue) - startPoint);
            var extensionWouldTakeUsPastDateTimeMaxValue = proposedExtension > maxPossibleExtension;

            if (extensionWouldTakeUsPastDateTimeMaxValue)
            {
                return Tuple.Create(DateTime.MaxValue, maxPossibleExtension);
            }
            else
            {
                return Tuple.Create(startPoint + proposedExtension, proposedExtension);
            }
        }
```

Reply:
Then add an overload to accept a Ticks-based StartDate. (We *could* do that calculation in PermitTest, but that method is already hella-complicated, so I'm inclined to avoid adding additional date-representation-faffing complexity to it.)

```
        private Tuple<DateTime, TimeSpan> ExtensionUpToMaxValue(long startPointTicks, TimeSpan proposedExtension, DateTime? maxDateTime = null)
        {
            var startDate = DateTime.MinValue + TimeSpan.FromTicks(startPointTicks);
            return ExtensionUpToMaxValue(startDate, proposedExtension, maxDateTime);
        }
```

Reply:
Then use that in Permit...Test:

```
var (proposedNewBlockDeadline,_) = ExtensionUpToMaxValue(blockedUntil_AtStartOfPermitHalfOpenCircuitTest, _currentBlockDuration);
```

Reply:
All good?

Reply:
Thanks @Brondahl for the factor out! I can only think of couple small things - can you push it? - easier to comment on pushed lines.  

Again, thanks for everything on this.

Reply:
Again, thanks for the :+1: factoring out of `ExtensionUpToMaxValue(...)`.  Comments:

Let's avoid `System.Tuple`; we don't need to use a reference type here. We can stick to `System.ValueTuple`, just using the `(foo, bar)` value-tuple syntax.

Second, `DateTime.MinValue + TimeSpan.FromTicks(startPointTicks)` -> `new DateTime(startPointTicks)`.  (Given both the consumers of `ExtensionUpToMaxValue(...)` actually master their data (`_blockedTill`) in `Ticks`, it might be cleaner just to reduce the two overloads `ExtensionUpToMaxValue` to a single overload `ExtensionUpToMaxValue(long startPointTicks, TimeSpan proposedExtension`, but I'm not fussed if you want to push past this and leave the two overloads for now.)

Reply:
>We can stick to System.ValueTuple
Ah, nice. Done.

>new DateTime(startPointTicks)
Ah, I looked for `DateTime.FromTicks()` but couldn't find it. Thank you :) Done.

I'm inclined to leave them separate. I think the rationale for the extension method is designed quite heavily around "we're representing DateTimes". If we switch to using Ticks everywhere, then we're *actually* using `long`s everywhere, and it will be tempting to start allowing anything up to `long.MaxValue` ... and then you'll end up with code to check that the `long` is a valid `DateTime` or `TimeSpan`.

Now that you've shown me the terse way to do Ticks => DateTime, I'd be happy to just have that inline in `PermitHalfOpenCircuitTest`, and drop the Ticks-based overload, so we only have the one, DateTime-based, method?

Reply:
Good points @Brondahl , nice changes :+1: